python-package/xgboost/spark/data.py

```diff
@@ -1,194 +1,181 @@
-    train_val_data = _prepare_train_val_data(
-        partition_data_iter, has_weight, has_validation, has_base_margin
-    )
-    if has_validation:
-        (
-            train_x,
-            train_y,
-            train_w,
-            train_b_m,
-            val_x,
-            val_y,
-            val_w,
-            val_b_m,
-        ) = train_val_data
-        training_dmatrix = DMatrix(
-            data=train_x,
-            label=train_y,
-            weight=train_w,
-            base_margin=train_b_m,
-            **dmatrix_kwargs,
-        )
-        val_dmatrix = DMatrix(
-            data=val_x,
-            label=val_y,
-            weight=val_w,
-            base_margin=val_b_m,
-            **dmatrix_kwargs,
+
+    train_data: Dict[str, List[np.ndarray]] = defaultdict(list)
+    valid_data: Dict[str, List[np.ndarray]] = defaultdict(list)
+
+    n_features: int = 0
+
+    def append_m(part: pd.DataFrame, name: str, is_valid: bool) -> None:
+        nonlocal n_features
+        if name in part.columns:
+            array = part[name]
+            if name == alias.data:
+                array = stack_series(array)
+                if n_features == 0:
+                    n_features = array.shape[1]
+                assert n_features == array.shape[1]
+
+            if is_valid:
+                valid_data[name].append(array)
+            else:
+                train_data[name].append(array)
+
+    def append_dqm(part: pd.DataFrame, name: str, is_valid: bool) -> None:
+        """Preprocessing for DeviceQuantileDMatrix"""
+        nonlocal n_features
+        if name == alias.data or name in part.columns:
+            if name == alias.data:
+                cname = feature_cols
+            else:
+                cname = name
+
+            array = part[cname]
+            if name == alias.data:
+                if n_features == 0:
+                    n_features = array.shape[1]
+                assert n_features == array.shape[1]
+
+            if is_valid:
+                valid_data[name].append(array)
+            else:
+                train_data[name].append(array)
+
+    def make(values: Dict[str, List[np.ndarray]], **kwargs: Any) -> DMatrix:
+        data = concat_or_none(values[alias.data])
+        label = concat_or_none(values.get(alias.label, None))
+        weight = concat_or_none(values.get(alias.weight, None))
+        margin = concat_or_none(values.get(alias.margin, None))
+        return DMatrix(
+            data=data, label=label, weight=weight, base_margin=margin, **kwargs
         )
-        return training_dmatrix, val_dmatrix
-
-    train_x, train_y, train_w, train_b_m = train_val_data
-    training_dmatrix = DMatrix(
-        data=train_x,
-        label=train_y,
-        weight=train_w,
-        base_margin=train_b_m,
-        **dmatrix_kwargs,
-    )
-    return training_dmatrix
+
+    is_dmatrix = feature_cols is None
```

Comment:
why use `DeviceQuantileDMatrix` only when feature_cols is not None ?

Reply:
Use DeviceQuantileDMatrix or not should be controlled by `use_gpu` param if I understand correctly?

Reply:
So far `QuantileDMatrix` is GPU only until https://github.com/dmlc/xgboost/pull/8050 is merged. `feature_cols` will also be a GPU-only thing for the recent future. (Actually I'm not entirely sure about the `feature_cols` parameter that's available on Jvm packages, I don't know how does it work with spark ml pipeline, I will leave these questions to @wbo4958 ).

Reply:
We don't need to support `feature_cols` param in xgboost pyspark estimator. The transformers in spark pipeline can assemble multiple feature columns into one vector type feature column.

Reply:
I think @wbo4958 wanted to avoid the vector assembler since xgboost needs to undo it. (the `stack_series`). He wanted to have the input data just like other python libraries which 1 column per feature. I think it's a reasonable optimization looking solely on XGBoost, but might not play well with spark ml pipelines.

Reply:
I will remove it for now.

Reply:
Actually, I will keep it for now and leave it as a followup.

Reply:
> but might not play well with spark ml pipelines.

It does not play well with both spark ml pipelines and other spark meta-estimators such as CrossValidator, OneVsRest, etc..

Reply:
Got it, I will remove it. cc @wbo4958 .

Reply:
@WeichenXu123, yeah, it indeed does not play well with some spark ml pipelines, but we can give users some warning about that. Some users may not use spark ml pipeline or meta-estimators at all, they may just want to train an XGBoost model. I think we should provide this.

Reply:
@wbo4958 
OK. But I doubt "1 column per feature" in spark dataframe and pass many columns to python UDF it might not have good performance when there're many features. We'd better benchmark it.

Reply:
@WeichenXu123, Yes, we will do some benchmark testing. BTW, may I ask why "pass many columns to python UDF it might not have good performance when there're many features", is the penalty happening on the final data size of ArrowRecordBatch?

Reply:
> is the penalty happening on the final data size of ArrowRecordBatch?

Yes, that's my concern. But I am not expert on this, maybe I am wrong. Let's benchmark.